protocol/0076-DANO-data-node.md

```diff
@@ -183,6 +183,21 @@ It must be possible to add to the data node APIs that return the result of calcu
 2. Standard nodes should retain data in accordance with the configured data retention policy (<a name="0076-DANO-027" href="#0076-DANO-027">0076-DANO-027</a>)
 3. Archival nodes should retain all data from the height at which they joined the network (<a name="0076-DANO-028" href="#0076-DANO-028">0076-DANO-028</a>)
 
+### API Request Rate Limiting
+
+1. Datanode should provide an optional mechanism for limiting the average number of requests per second over on its API
+2. That rate should be specified in the datanode configuration file
+3. A client may, over a short period of time, make requests at a greater frequency than the limit as long as the average rate over a longer period of time is not exceeded. (<a name="0076-DANO-029" href="#0076-DANO-029">0076-DANO-029</a>)
+4. The extent to which clients may of 'burst' requests should also be capped and specified in the datanode configuration file (<a name="0076-DANO-030" href="#0076-DANO-030">0076-DANO-030</a>)
+5. Limits should be enforced on a per-client basis. Source IP address is a sufficient discriminator (<a name="0076-DANO-031" href="#0076-DANO-031">0076-DANO-031</a>)
+6. Headers or metadata should be included in each API response indicating to the client what the limits are, and how close they currently are to exceeding them (<a name="0076-DANO-032" href="#0076-DANO-032">0076-DANO-032</a>)
+7. If limits are exceeded an API appropriate error response should be returned, containing similar headers or metadata (<a name="0076-DANO-033" href="#0076-DANO-033">0076-DANO-033</a>)
+8. If the client continues to make requests that generate error responses due to rate limiting, eventually the client should banned (<a name="0076-DANO-034" href="#0076-DANO-034">0076-DANO-034</a>)
```

Comment:
`eventually the client should banned` - the trigger point for the banning should be defined for the AC to be testable.

Reply:
OK. There is a defined mechanism for this (a second bucket of 'naughty' tokens of the same size/refill rate, from which a token is removed on every unsuccessful-due-to-rate-limiting request). I'll was slightly reluctant to put that in the spec as I wasn't sure that was going to be the final solution but you're right - to test it you need to know so I'll update it.

Reply:
I've changed the definition of point 8 to be a bit more concrete. I wanted to keep the specification kind of short and snappy so didn't go into massive detail, if you want me to lay out how it actually works I can.